### PR TITLE
Abstract events into an event listener

### PR DESCRIPTION
**Pull request by kfichter**
Abstracts all events into `RootEventListener`, which emits events only once they're "finalized" in the root chain. Should avoid a lot of issues with reorgs. Needs to be integration tested, waiting on testing lang.

### CODE REVIEW

File: plasma/child_chain/child_chain.py [hunk 1]
```diff
@@ -20,12 +22,22 @@ def __init__(self, authority, root_chain):
         self.current_block = Block()
         self.pending_transactions = []
 
-        # Register for deposit event listener
-        deposit_filter = self.root_chain.on('Deposit')
-        deposit_filter.watch(self.apply_deposit)
+        self.event_listener = RootEventListener(root_chain, finality=1)
```
Comment (DavidKnott) [inline plasma/child_chain/child_chain.py]: How do you feel about changing `finality` to `confirmations` as it's more descriptive?

File: plasma/child_chain/root_event_listener.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        Args:
+            event_name (str): Name of event to watch.
+        """
+
+        while event_name in self.active_events:
+            current_block = self.w3.eth.getBlock('latest')
+            event_filter = self.root_chain.eventFilter(event_name, {
+                'fromBlock': current_block['number'] - (self.finality * 2),
+                'toBlock': current_block['number'] - self.finality
+            })
```
Comment (DavidKnott) [inline plasma/child_chain/root_event_listener.py]: How did you decide upon the (-finality, -finality * 2) gap for the filter?  Do you think if we just set it to -finality events might be missed?

File: tests/child_chain/test_child_chain.py [hunk 3]
```diff
@@ -32,7 +33,9 @@ def child_chain():
     # Create a block with those transactions
     child_chain.blocks[1] = Block([tx1, tx2])
 
-    return child_chain
+    yield child_chain
+
+    child_chain.event_listener.stop_all()
```
Comment (DavidKnott) [inline tests/child_chain/test_child_chain.py]: Shouldn't stopping the events that the child chain is instantiated with come before `yield`?

File: tests/child_chain/test_child_chain.py [hunk 4]
```diff
@@ -141,3 +144,25 @@ def test_apply_deposit(child_chain):
...
+        'transactionIndex': 0,
+        'transactionHash': '0x35e6446818b53b2c4537ebba32b9453b274286ffbb25e5b521a6b0a33e2cb953',
+        'address': '0xA3B2a1804203b75b494028966C0f62e677447A39',
+        'blockHash': '0x2550290dd333ea2876539b7ba474a804a9143b0d4ecb57b9d824f07ffd016747',
+        'blockNumber': 1
+    }
+    child_chain.apply_exit(sample_event)
+
+    # Transaction now marked spent
+    assert child_chain.blocks[blknum].transaction_set[txindex].spent1
```
Comment (DavidKnott) [inline tests/child_chain/test_child_chain.py]: Can you add an assert verifying that the output that's now spent was not before `apply_exit`?

File: plasma/child_chain/child_chain.py (hunk 1 above)
Comment (kfichter) [reply plasma/child_chain/child_chain.py]: Agreed.

File: plasma/child_chain/root_event_listener.py (hunk 2 above)
Comment (kfichter) [reply plasma/child_chain/root_event_listener.py]: I thought events might be missed if the client lost connection temporarily. Maybe it's better to just update some internal variable that stores the latest block the client has polled instead.

File: tests/child_chain/test_child_chain.py (hunk 3 above)
Comment (kfichter) [reply tests/child_chain/test_child_chain.py]: `yield` returns the child chain immediately (for use in tests), which starts event listeners. The `stop_all()` statement is just teardown after the tests have completed.

**Pull request merged**